GREYErrorFormatter.m

```diff
@@ -0,0 +1,136 @@
+//
+// Copyright 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import "GREYErrorFormatter.h"
+
+#import "GREYObjectFormatter.h"
+#import "GREYInteraction.h"
+#import "GREYErrorConstants.h"
+#import "NSError+GREYCommon.h"
+#import "GREYError+Private.h"
+
+@interface GREYErrorFormatter ()
+
+@property(readonly, nonatomic) GREYError *error;
+
+@end
+
+@implementation GREYErrorFormatter
+
+#pragma mark - Init
+
+- (instancetype)initWithError:(GREYError *)error {
+  self = [super init];
+  if (self) {
+    _error = error;
+  }
+  return self;
+}
+
+#pragma mark - Public Methods
+
+- (NSString *)formattedDescription {
+  if (GREYShouldUseErrorFormatterForError(_error)) {
+    return [self elementNotFoundDescription];
+  }
+  return [GREYObjectFormatter formatDictionary:[_error grey_descriptionDictionary]
+                                        indent:kGREYObjectFormatIndent
+                                     hideEmpty:YES
+                                      keyOrder:nil];
+}
+
+#pragma mark - Private Methods
+
+- (NSString *)elementNotFoundDescription {
```

Comment:
This method name is too tightly coupled to the error being described. For new error codes, this would mean we would have to have a function for each error code. 

I think a better way would be if we print any of the below information if it is provided to us and leave it to the error creator to actually fill out the error. WDYT?

Reply:
Updated. I first look up the set of keys that should be shown for a given code/domain, and then pass that to another method that prints out the values of those keys.